{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "7d0951ea",
   "metadata": {},
   "source": [
    "# Rosalind ORF\n",
    "\n",
    "**Given**: A DNA sequence in FASTA format\n",
    "\n",
    "**Asked:** all possible peptides that can be translated from it\n",
    "\n",
    "## The Plan\n",
    "\n",
    "- [x] transcribe the DNA to RNA (RNA exercise)\n",
    "- [x] find all reading frames $\\rightarrow$ three forward, three reverse $\\rightarrow$ REVC exercise\n",
    "- [x] find valid ORFs in an RNA sequence <font color=\"red\">!!!!</font>\n",
    "- [x] translate the mRNAs to peptides (PROT exercise)\n",
    "- [x] make sure there are no duplicate protein sequences!\n",
    "\n",
    "Ideas:\n",
    "\n",
    "- much like in PROT, we might need to work with triplets\n",
    "- whatever we do for a single reading frame, we will have to repeat 5x more $\\rightarrow$ functions?\n",
    "    - function that takes RNA sequence and produces all valid peptides"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "548e92f3",
   "metadata": {},
   "source": [
    "## Set up environment\n",
    "\n",
    "- [x] load helper functions\n",
    "- [x] get testing data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "08fdaec0",
   "metadata": {},
   "outputs": [],
   "source": [
    "from util import read_fasta, reverse_complement\n",
    "from util import codons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e29a0422",
   "metadata": {},
   "outputs": [],
   "source": [
    "fasta_file = read_fasta('test.fa')\n",
    "fasta = list(fasta_file.values())[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "7e366d23",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'AGCCATGTAGCTAACTCAGGTTACATGGGGATGACCCCGCGACTTGGATTAGAGTCTCTTTTGGAATAAGCCTGAATGATCCGAGTAGCATCTCAG'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b9832caa",
   "metadata": {},
   "source": [
    "### 1. transcribe to RNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "93b842e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna = fasta.replace('T', 'U')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cb5bc909",
   "metadata": {},
   "source": [
    "### 2. find all reading frames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "49faa270",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_reading_frames(rna):\n",
    "    # three forward, three backward reading frames\n",
    "    rf1 = rna\n",
    "    rf2 = rna[1:]\n",
    "    rf3 = rna[2:]\n",
    "\n",
    "    revc = reverse_complement(rna)\n",
    "    rf4 = revc\n",
    "    rf5 = revc[1:]\n",
    "    rf6 = revc[2:]\n",
    "    return [rf1, rf2, rf3, rf4, rf5, rf6]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "1f1719d1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['AGCCAUGUAGCUAACUCAGGUUACAUGGGGAUGACCCCGCGACUUGGAUUAGAGUCUCUUUUGGAAUAAGCCUGAAUGAUCCGAGUAGCAUCUCAG',\n",
       " 'GCCAUGUAGCUAACUCAGGUUACAUGGGGAUGACCCCGCGACUUGGAUUAGAGUCUCUUUUGGAAUAAGCCUGAAUGAUCCGAGUAGCAUCUCAG',\n",
       " 'CCAUGUAGCUAACUCAGGUUACAUGGGGAUGACCCCGCGACUUGGAUUAGAGUCUCUUUUGGAAUAAGCCUGAAUGAUCCGAGUAGCAUCUCAG',\n",
       " 'CUGAGAUGCUACUCGGAUCAUUCAGGCUUAUUCCAAAAGAGACUCUAAUCCAAGUCGCGGGGUCAUCCCCAUGUAACCUGAGUUAGCUACAUGGCU',\n",
       " 'UGAGAUGCUACUCGGAUCAUUCAGGCUUAUUCCAAAAGAGACUCUAAUCCAAGUCGCGGGGUCAUCCCCAUGUAACCUGAGUUAGCUACAUGGCU',\n",
       " 'GAGAUGCUACUCGGAUCAUUCAGGCUUAUUCCAAAAGAGACUCUAAUCCAAGUCGCGGGGUCAUCCCCAUGUAACCUGAGUUAGCUACAUGGCU']"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "frames = get_reading_frames(rna)\n",
    "frames"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f6f2540e",
   "metadata": {},
   "source": [
    "## 3. finding ORFs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "5f55370f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# just to practice, use RF1\n",
    "rf = frames[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b6d4aff4",
   "metadata": {},
   "source": [
    "I notice that I cannot solve this in one go; the problem are overlapping peptides. What is the\n",
    "problem  about them? They share a stop codon, but have different start codons $\\rightarrow$ there\n",
    "is no solution with just one for loop.\n",
    "\n",
    "My idea:\n",
    "\n",
    "- [x] find all start codons\n",
    "- [x] given a start codon, find the correspodnign ORF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "12c76519",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_AUG(rf):\n",
    "    starting_positions = []\n",
    "    for i in range(0, len(rf), 3):\n",
    "        codon = rf[i:i+3]\n",
    "        if codon == 'AUG':\n",
    "            starting_positions.append(i)\n",
    "    return starting_positions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "b4d26b5d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[24, 30, 75]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "starts = find_AUG(rf)\n",
    "starts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "2995166d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_ORF_and_translate(rf, start, codons):\n",
    "    peptide = ''\n",
    "    for i in range(start, len(rf), 3):\n",
    "        codon = rf[i:i+3]\n",
    "        if codons[codon] is None:\n",
    "            return peptide\n",
    "        peptide += codons[codon]\n",
    "    return None"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "97965ce1",
   "metadata": {},
   "source": [
    "negative test: this sequence contains a start codon but no end codon; our function should return\n",
    "None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "16a6c185",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "None\n"
     ]
    }
   ],
   "source": [
    "neg = 'AUGGGGAUGACCCCGCGACUUGGAUUAGAGUCUCUUUUGGAA'\n",
    "peptide = find_ORF_and_translate(neg, 0, codons)\n",
    "print(peptide)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "29350d45",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_all_proteins(rf, codons):\n",
    "    starts = find_AUG(rf)\n",
    "    proteins = []\n",
    "    for start in starts:\n",
    "        peptide = find_ORF_and_translate(rf, start, codons)\n",
    "        if peptide is not None:\n",
    "            proteins.append(peptide)\n",
    "    return proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "3d06d52c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['MGMTPRLGLESLLE', 'MTPRLGLESLLE']"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "find_all_proteins(frames[0], codons)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d9b33551",
   "metadata": {},
   "source": [
    "## Putting it all together:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "eb41488f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def all_possible_proteins(fasta_path, codons):\n",
    "    # 1. read the file\n",
    "    fasta_file = read_fasta(fasta_path)\n",
    "    fasta = list(fasta_file.values())[0]\n",
    "    # 2. transcribe to RNA\n",
    "    rna = fasta.replace('T', 'U')\n",
    "    # 3. get reading frames\n",
    "    frames = get_reading_frames(rna)\n",
    "    proteins = []\n",
    "    for frame in frames:\n",
    "        to_add = find_all_proteins(frame, codons)\n",
    "        # make sure the new proteins are not\n",
    "        # already in our list\n",
    "        for prot in to_add:\n",
    "            if prot not in proteins:\n",
    "                proteins.extend([prot])\n",
    "\n",
    "    print('\\n'.join(proteins))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de35e82f",
   "metadata": {},
   "source": [
    "## Solve example"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "3e1ae405",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MGMTPRLGLESLLE\n",
      "MTPRLGLESLLE\n",
      "M\n",
      "MLLGSFRLIPKETLIQVAGSSPCNLS\n"
     ]
    }
   ],
   "source": [
    "all_possible_proteins('test.fa', codons)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f8aec27",
   "metadata": {},
   "source": [
    "## Solve exercise"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "22fa7bb2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MGAIGTMRDKAHNSFQSYAWLYGYYNDRCTRALTKPYELLFERVKNTHQHLILCSVSYSSPILSLDTSCCNR\n",
      "MRDKAHNSFQSYAWLYGYYNDRCTRALTKPYELLFERVKNTHQHLILCSVSYSSPILSLDTSCCNR\n",
      "MPVPLIGAAMAFETQPR\n",
      "MAFETQPR\n",
      "MSCMGPESVQIV\n",
      "MGPESVQIV\n",
      "MIYNSFLRAIKRSKRRDFCHSYVEYARCRTFSRQALAVFLTSTFPQEHHRPEVYIES\n",
      "MLGYTVITTTDVLER\n",
      "MNRPFYEFLGRPLR\n",
      "MALRQDWSTCRCP\n",
      "MPLFSPIS\n",
      "MGLLGIIFSYL\n",
      "MRAAVRLAGKL\n",
      "MPQSVKNGHHN\n",
      "MY\n",
      "MGRPLTLDVYFWSMMFLRES\n",
      "MMFLRES\n",
      "MFLRES\n",
      "MTEVAPF\n",
      "MMPSNPMRAAARLKVPYRALPRLLKWSG\n",
      "MPSNPMRAAARLKVPYRALPRLLKWSG\n",
      "MRAAARLKVPYRALPRLLKWSG\n",
      "MGENNGMAFETQPR\n",
      "MYTSGR\n",
      "MAWRSKHSPVS\n",
      "MWTSPA\n",
      "MFVT\n",
      "MVGSSLIEY\n",
      "MMSVFNTLGHGAPSNSRCILLVDDVPAGKLKSGIPLELAC\n",
      "MSVFNTLGHGAPSNSRCILLVDDVPAGKLKSGIPLELAC\n",
      "MNDRSRAFLNAL\n",
      "MVGLRDGRKQWHGVRNTAPLANGAVFRTP\n",
      "MYPATV\n"
     ]
    }
   ],
   "source": [
    "all_possible_proteins('/Users/npapadop/Downloads/rosalind_orf.txt', codons)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "plotting",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
