{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b4966b9e-0bcf-4181-b948-82f6dc2ea9bf",
   "metadata": {},
   "source": [
    "# Translating RNA to protein (rosalind-PROT)\n",
    "\n",
    "The 20 commonly occurring amino acids are abbreviated by using 20 letters from the English alphabet (all letters except for B, J, O, U, X, and Z). Protein strings are constructed from these 20 symbols. Henceforth, the term genetic string will incorporate protein strings along with DNA strings and RNA strings.\n",
    "\n",
    "The RNA codon table dictates the details regarding the encoding of specific codons into the amino acid alphabet.\n",
    "\n",
    "**Given:** An RNA string s corresponding to a strand of mRNA (of length at most 10 kbp).\n",
    "\n",
    "**Return:** The protein string encoded by s.\n",
    "\n",
    "### Sample Dataset\n",
    "\n",
    "```\n",
    "AUGGCCAUGGCGCCCAGAACUGAGAUCAAUAGUACCCGUAUUAACGGGUGA\n",
    "```\n",
    "\n",
    "### Sample Output\n",
    "\n",
    "```\n",
    "MAMAPRTEINSTRING\n",
    "```\n",
    "\n",
    "## Solution:\n",
    "\n",
    "The task is to translate an RNA sequence to aminoacids; and clearly, we are given an open reading frame, so there is no need to go searching for start codons. What we need to solve this are\n",
    "\n",
    "1. a way to split up the RNA sequence to triplets\n",
    "2. a way to map triplets of nucleotides to a single aminoacid\n",
    "\n",
    "### 1. splitting the RNA to triplets\n",
    "\n",
    "Let's first try to solve the first part, splitting the RNA sequence to triplets. I will be using the example sequence to practice."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9babd9dd-9b2a-4bbe-b622-e8699faa32a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "s = \"AUGGCCAUGGCGCCCAGAACUGAGAUCAAUAGUACCCGUAUUAACGGGUGA\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1c390ea-fe1f-4c16-9e7f-a1817bca02fb",
   "metadata": {},
   "source": [
    "What is the first codon?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "66445386-783e-4f0c-a281-a68f9e9a7fa6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'AUG'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "s[0:3]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c892f1ca-82f4-4608-8d0e-5cca00e7c27b",
   "metadata": {},
   "source": [
    "I can verify that manually: the first codon is indeed from indices 0 to 3. What is the second codon?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "bf06db02-ceaf-4053-bdfc-55c15a24c634",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'GCC'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "s[3:6]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d64f1dca-dde9-4100-a99f-f13d420985cf",
   "metadata": {},
   "source": [
    "Again, I can verify this by eye. How would I get the i'th codon?\n",
    "\n",
    "$\\rightarrow$ this might be too hard to imagine immediately, so let's try writing everything down:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7d4792a6-d303-43e8-a647-e90e345cebd0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'AUG'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "s[0:3] # first codon, or 0th, for Python\n",
    "s[3:6] # second codon, or 1st for Python\n",
    "s[6:9] # third codon, or 2nd for Python"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ae22066-a9ee-4a3f-8bab-29e61b78b481",
   "metadata": {},
   "source": [
    "The 10th codon will have 9 codons before it; since every codon is 3nt long, this means there will have been $3 \\times 9 = 27$ nucleotides before we arrive at the 10th codon. The 10th codon will then also stretch for 3 nucleotides.\n",
    "\n",
    "This means that for the i'th codon there will be $i-1$ codons before it; and therefore $(i-1) \\times 3$ nucleotides. If we start at 0, like Python does, we can modify that to $i$ codons and therefore use $i \\times 3$ as the starting position of the i-th codon:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e402f897-8d7f-4c4c-8e36-a6ffb25fff19",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'GCC'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "i = 1 # in Python notation\n",
    "start = i * 3\n",
    "end = start + 3\n",
    "s[start:end]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0a4edbbf-3e57-419f-bc80-185be4ed7e5c",
   "metadata": {},
   "source": [
    "Then $i$ represents the current codon; how many of those will be contained in an ORF? $\\rightarrow$ of course the length of the sequence divided by three, as the sequence is made up of triplets."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "9096761f-f258-42f1-b0b2-90bac8881337",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AUG\n",
      "GCC\n",
      "AUG\n",
      "GCG\n",
      "CCC\n",
      "AGA\n",
      "ACU\n",
      "GAG\n",
      "AUC\n",
      "AAU\n",
      "AGU\n",
      "ACC\n",
      "CGU\n",
      "AUU\n",
      "AAC\n",
      "GGG\n",
      "UGA\n"
     ]
    }
   ],
   "source": [
    "no_codons = len(s) // 3 # integer division\n",
    "for i in range(no_codons):\n",
    "    start = i * 3\n",
    "    end = start + 3\n",
    "    codon = s[start:end]\n",
    "    print(codon)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01891047-578a-4ad8-a689-af74c9ae04fd",
   "metadata": {},
   "source": [
    "Looking at the sequence verifies that we did this correctly!\n",
    "\n",
    "Of course, now that we get how this works, we can also do it with slicing notation, exploiting the fact that _every third position_ in the RNA string will be a codon start position:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "f6308d2f-0e33-473f-97db-e1714c8731c0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AUG\n",
      "GCC\n",
      "AUG\n",
      "GCG\n",
      "CCC\n",
      "AGA\n",
      "ACU\n",
      "GAG\n",
      "AUC\n",
      "AAU\n",
      "AGU\n",
      "ACC\n",
      "CGU\n",
      "AUU\n",
      "AAC\n",
      "GGG\n",
      "UGA\n"
     ]
    }
   ],
   "source": [
    "for i in range(0, len(s), 3): # the third argument to range is the step size!\n",
    "    # start = i\n",
    "    # end = start + 3\n",
    "    # codon = s[start:end]\n",
    "    codon = s[i:i+3] # both options are equally correct; this is just more compact!\n",
    "    print(codon)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76a9ba5c-e3a9-40c0-a29c-a6c139d37022",
   "metadata": {},
   "source": [
    "### 2. replacing codons with their respective aminoacids\n",
    "\n",
    "Following the link to the codon table we see that each codon is paired up with one aminoacid. We clearly have pairs of things that belong together; maybe a dictionary would be a natural way of saving this information.\n",
    "\n",
    "But if it's a dictionary, what are the values and what are the keys? We know from the definition of dictionaries that the keys have to be unique. Each aminoacid can be encoded by multiple codons, but each codon encodes exactly one aminoacid. Therefore, the keys should be codons, and the values aminoacids.\n",
    "\n",
    "We can copy the codon table from the exercise, but we need to be careful: this dictionary assigns an one-letter aminoacid to each codon. This means that the stop codons will need to be handled in a special manner, since \"Stop\" is a four-letter word, breaking the symmetry of the rest of the table. We will leave them like that for now, but may need to revisit this later."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "50d96d67-52a3-4f4c-b580-b6d70ac7e727",
   "metadata": {},
   "outputs": [],
   "source": [
    "codons = {\n",
    "    \"UUU\": \"F\", \"CUU\": \"L\", \"AUU\": \"I\", \"GUU\": \"V\",\n",
    "    \"UUC\": \"F\", \"CUC\": \"L\", \"AUC\": \"I\", \"GUC\": \"V\",\n",
    "    \"UUA\": \"L\", \"CUA\": \"L\", \"AUA\": \"I\", \"GUA\": \"V\",\n",
    "    \"UUG\": \"L\", \"CUG\": \"L\", \"AUG\": \"M\", \"GUG\": \"V\",\n",
    "    \"UCU\": \"S\", \"CCU\": \"P\", \"ACU\": \"T\", \"GCU\": \"A\",\n",
    "    \"UCC\": \"S\", \"CCC\": \"P\", \"ACC\": \"T\", \"GCC\": \"A\",\n",
    "    \"UCA\": \"S\", \"CCA\": \"P\", \"ACA\": \"T\", \"GCA\": \"A\",\n",
    "    \"UCG\": \"S\", \"CCG\": \"P\", \"ACG\": \"T\", \"GCG\": \"A\",\n",
    "    \"UAU\": \"Y\", \"CAU\": \"H\", \"AAU\": \"N\", \"GAU\": \"D\",\n",
    "    \"UAC\": \"Y\", \"CAC\": \"H\", \"AAC\": \"N\", \"GAC\": \"D\",\n",
    "    \"UAA\": \"Stop\", \"CAA\": \"Q\", \"AAA\": \"K\", \"GAA\": \"E\",\n",
    "    \"UAG\": \"Stop\", \"CAG\": \"Q\", \"AAG\": \"K\", \"GAG\": \"E\",\n",
    "    \"UGU\": \"C\", \"CGU\": \"R\", \"AGU\": \"S\", \"GGU\": \"G\",\n",
    "    \"UGC\": \"C\", \"CGC\": \"R\", \"AGC\": \"S\", \"GGC\": \"G\",\n",
    "    \"UGA\": \"Stop\", \"CGA\": \"R\", \"AGA\": \"R\", \"GGA\": \"G\",\n",
    "    \"UGG\": \"W\", \"CGG\": \"R\", \"AGG\": \"R\", \"GGG\": \"G\"\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec6936d9-5218-4fab-bf5d-e19e77aeb0eb",
   "metadata": {},
   "source": [
    "Now we can simply go through our codons and write out the correct aminoacid for each codon!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "0ac4c6d8-2a6e-4ad0-9604-3cb376a805b3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "M\n",
      "A\n",
      "M\n",
      "A\n",
      "P\n",
      "R\n",
      "T\n",
      "E\n",
      "I\n",
      "N\n",
      "S\n",
      "T\n",
      "R\n",
      "I\n",
      "N\n",
      "G\n",
      "Stop\n"
     ]
    }
   ],
   "source": [
    "for i in range(0, len(s), 3):\n",
    "    codon = s[i:i+3]\n",
    "    aminoacid = codons[codon]\n",
    "    print(aminoacid)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "80053c6e-dae3-4ff5-a150-30eebf8b9919",
   "metadata": {},
   "source": [
    "We could try making this into a word, but the \"Stop\" at the end will mess things up. Maybe we should check for this somehow."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "f89ae863-25e9-420d-998e-0edfa80b16a3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MAMAPRTEINSTRING\n"
     ]
    }
   ],
   "source": [
    "peptide = \"\"\n",
    "for i in range(0, len(s), 3):\n",
    "    codon = s[i:i+3]\n",
    "    aminoacid = codons[codon]\n",
    "    if aminoacid != \"Stop\":\n",
    "        peptide = peptide + aminoacid\n",
    "print(peptide)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d68a61a2-99e8-4dd6-b24e-f9f07577510e",
   "metadata": {},
   "source": [
    "We're done!\n",
    "\n",
    "Please note: this approach will not stop reading an RNA string if a stop codon is seen; it will continue until there is no more sequence to parse."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
